File: game_logic/player.py
class Player:


    def __init__(self, path_lengths: list[int], name: str) -> None:
        self.total_moves: int = 0
        self.state: list[int] = path_lengths.copy()
        self.temp: dict[int, int] = {}
        self.name = name


    def take_action(self, path: int) -> None:
        if path not in self.temp:
            self.temp[path] = 1
        else:
            self.temp[path] += 1
        return
    

    def undo_action(self, path: int) -> None:
        if self.temp[path] == 1:
            del self.temp[path]
        else:
            self.temp[path] -= 1


    def update_state(self) -> None:
        for path in self.temp:
            self.state[path] -= self.temp[path]
        return


    def reset_temp(self) -> None:
        self.temp.clear()
        return


    def add_moves(self, move: int) -> None:
        self.total_moves += move


    def is_winner(self, goal: int) -> bool:
        count = 0
        for x in self.state:
            if x == 0:
                count += 1
        return count >= goal

File: player.js
// class Player {
//   constructor(client) {
//     this.id = client.id;
//     this.client = client;
//     this.totalMoves = 0;
//     this.state = [];
//     this.temp = [];
//   }

//   initialize(pathLengths) {
//     this.state = [...pathLengths];
//     this.temp = new Array(pathLengths.length).fill(0);
//   }

//   numTemp() {
//     let count = 0;
//     for (const x of this.temp) {
//       if (x > 0) {
//         count++;
//       }
//     }
//     return count;
//   }

//   takeAction(path) {
//     this.temp[path]++;
//   }

//   undoAction(path) {
//     this.temp[path]--;
//   }

//   updateState() {
//     for (const [path, progress] of this.temp.entries()) {
//       this.state[path] -= progress;
//     }
//   }

//   resetTemp() {
//     this.temp.fill(0);
//   }

//   addMoves(move) {
//     this.totalMoves += move;
//   }

//   score() {
//     let count = 0;
//     for (let x of this.state) {
//       if (x == 0) {
//         count++;
//       }
//     }
//     return count;
//   }

//   isWinner(goal) {
//     return this.score() >= goal;
//   }
// }

// export { Player };
